sycl/source/detail/context_impl.cpp

```diff
@@ -115,6 +115,25 @@ cl_context context_impl::get() const {
 bool context_impl::is_host() const { return MHostContext; }
 
 context_impl::~context_impl() {
+  // In case a user is leaking a memory object we may still have attached
+  // resources. These resources will be released with the context, but we need
+  // to do it before we release the backend context to avoid confusing errors.
+  std::unordered_map<const SYCLMemObjI *,
+                     std::vector<std::shared_ptr<const void>>>
+      MemObjLTARes;
+  {
+    std::lock_guard<std::mutex> lock(MMemObjLifetimeAttachedResourcesMutex);
```

Comment:
Is there any reason to do swapping here to reduce critical sections? I would expect that this function can be executed by one thread by definition and no other threads are supposed to access `context_impl` resources during it's destruction. Suggest just doing two `clear`'s under the lock to simplify the code.

Reply:
I was considering this and you are right that no threads should be conflicting with this, but since it is not a recursive lock we could have an attached buffer try to delete and detach its resources, which would cause a dead-lock here as it tries to lock this to check if it has any attached resources.